pynumdiff/total_variation_regularization/_total_variation_regularization.py

```diff
@@ -1,123 +1,112 @@
+                            In testing, 'MOSEK' was the most robust.
+  
+    :return: tuple[np.array, np.array] of\n
+             - **x_hat** -- estimated (smoothed) x
+             - **dxdt_hat** -- estimated derivative of x
     """
-
     # Normalize
     mean = np.mean(x)
     std = np.std(x)
     x = (x-mean)/std
 
     # Define the variables for the highest order derivative and the integration constants
-    var = cvxpy.Variable(len(x) + N)
+    deriv_values = cvxpy.Variable(len(x)) # values of the order^th derivative, in which we're penalizing variation
+    integration_constants = cvxpy.Variable(order) # constants of integration that help get us back to x
 
     # Recursively integrate the highest order derivative to get back to the position
-    derivatives = [var[N:]]
-    for i in range(N):
-        d = cvxpy.cumsum(derivatives[-1]) + var[i]
-        derivatives.append(d)
-
-    # Compare the recursively integration position to the noisy position
-    sum_squared_error = cvxpy.sum_squares(derivatives[-1] - x)
-
-    # Total variation regularization on the highest order derivative
-    r = cvxpy.sum(gamma*cvxpy.tv(derivatives[0]))
+    y = deriv_values
+    for i in range(order):
+        y = cvxpy.cumsum(y) + integration_constants[i]
 
     # Set up and solve the optimization problem
-    obj = cvxpy.Minimize(sum_squared_error + r)
-    prob = cvxpy.Problem(obj)
+    prob = cvxpy.Problem(cvxpy.Minimize(
+        # Compare the recursively integrated position to the noisy position, and add TVR penalty
+        cvxpy.sum_squares(y - x) + cvxpy.sum(gamma*cvxpy.tv(deriv_values)) ))
     prob.solve(solver=solver)
 
-    # Recursively calculate the value of each derivative
-    final_derivative = var.value[N:]
-    derivative_values = [final_derivative]
-    for i in range(N):
-        d = np.cumsum(derivative_values[-1]) + var.value[i]
-        derivative_values.append(d)
-    for i, _ in enumerate(derivative_values):
-        derivative_values[i] = derivative_values[i]/(dt**(N-i))
+    # Recursively integrate the final derivative values to get back to the function and derivative values
+    y = deriv_values.value
+    for i in range(order-1): # stop one short to get the first derivative
+        y = np.cumsum(y) + integration_constants.value[i]
+    dxdt_hat = y/dt # y only holds the dx values; to get deriv scale by dt
+    x_hat = np.cumsum(y) + integration_constants.value[order-1] # smoothed data
 
-    # Extract the velocity and smoothed position
-    dxdt_hat = derivative_values[-2]
-    x_hat = derivative_values[-1]
-
-    dxdt_hat = (dxdt_hat[0:-1] + dxdt_hat[1:])/2
+    dxdt_hat = (dxdt_hat[0:-1] + dxdt_hat[1:])/2 # take first order FD to smooth a touch
     ddxdt_hat_f = dxdt_hat[-1] - dxdt_hat[-2]
-    dxdt_hat_f = dxdt_hat[-1] + ddxdt_hat_f
+    dxdt_hat_f = dxdt_hat[-1] + ddxdt_hat_f # What is this doing? Could we use a 2nd order FD above natively?
     dxdt_hat = np.hstack((dxdt_hat, dxdt_hat_f))
 
     # fix first point
```

Comment:
Does the first point need to be fixed this way? I think you're just losing information with this update to the `[0]` index.